### DISCUSSION TITLE
Verdaccio Roadmap

### DISCUSSION THREAD

**Discussion by juanpicado** [verdaccio/verdaccio#1690] [General]
Hi, I'm @juanpicado and the intent of this story (epic) is to describe what we would like to have for Verdaccio next major.

https://github.com/verdaccio/verdaccio/tree/master

next major milestone 

https://github.com/verdaccio/verdaccio/milestone/22

```bash
docker pull verdaccio/verdaccio:nightly-master
```
or
```bash
npm install --global verdaccio@6-next
```
> ⚠️ don't use any of those images for production, are unstable and change quickly over time.

This year we will have a longer release, the project has grown and with a long tail of users, and at this point, we did not want to introduce big breaking changes in order to keep the backward compatibility. 
Verdaccio next major release will keep the nature of the predecessors, simple, zero-configuration need it, lightweight and open-source but with **a few breaking changes**.

The project needs to evolve and might break some plugins or configurations, so in order to achieve a calm transition *beta* and *alpha* will be longer, as usual, better migration guides and so on, not to mention we are doing this on **our spare time**.  We will define a final date release when we are sure we have enough topics closed. Verdaccio 5 still will be the major release and maintained as usual.  

As a reminder, [we accept donations](https://opencollective.com/verdaccio), this helps keep motivation up and move things faster.

If you want to contribute, feel free to open a discussion about any of these topics, or something else you think would be relevant, do not hesitate to reach me out.

### Structure

In the past was structured in several repositories, currently for the next major all the source code is hosted on this repository.

* [Next Release](https://github.com/verdaccio/verdaccio/discussions/2970)
* [Next Major Milestone · GitHub](https://github.com/verdaccio/verdaccio/milestone/22)
* [Projects In Progress· GitHub](https://github.com/verdaccio/verdaccio/projects) 

### Source code cleanup

A clean code base is a signal of it health, there is a lot of pending task, lint, callback hell, replace dependencies by native code, etc:

* [ ] fastify integration #2155 #2623  @dianmorales wip
* [x] [feat: migrate yarn workspaces by juanpicado · Pull Request #1546 · verdaccio/verdaccio · GitHub](https://github.com/verdaccio/verdaccio/pull/1546)
* [x] ~Apply latest `@verdaccio/babel-preset` and `@verdaccio/eslint-config`, `@verdaccio/types`.~ 
* [x] [clean up eslint warnings ·
    - [ ] clean all warnings on master `pnpm lint` and update `--max-warnings 49 ` script.
   -  [x]  Issue #46 · verdaccio/monorepo · GitHub](https://github.com/verdaccio/monorepo/issues/46)
* [x] ~[feat: use readable-stream 3 · Issue #32 · verdaccio/monorepo · GitHub](https://github.com/verdaccio/monorepo/issues/32)~
* [ ] Replace request by "xxxx" https://github.com/verdaccio/verdaccio/issues/1807
* [ ] Promisify Storage, Auth and Local Storage (this might also force also promisify readme, file-lock, local-storage plugin), in other words encorage `async/await` instead callbacks.  #1225 
* [ ] Replace dependencies by native functionalities provided by Node.js (less dependencies, the better) eg: we might use small method of a dependency, it could be part of our utils, this will help with maintenance.
*
#### Plugins

The Plugin support, the main pain comes from there, here a resume of the tickets we would like to address in the major release.

* [ ] [Multiple Authentication Plugins · Issue #1442 · verdaccio/verdaccio · GitHub](https://github.com/verdaccio/verdaccio/issues/1442)
* [ ] [Forcibly republish the package with the same version · Issue #1402 · verdaccio/verdaccio · GitHub](https://github.com/verdaccio/verdaccio/issues/1402)
* [x] [Plugin System is weird · Issue #1394 · verdaccio/verdaccio · GitHub](https://github.com/verdaccio/verdaccio/issues/1394) ✅ https://github.com/verdaccio/verdaccio/pull/3370
* [x] [Collisions between main verdaccio and plugins configuration · Issue #928 · verdaccio/verdaccio · GitHub](https://github.com/verdaccio/verdaccio/issues/928) ✅ https://github.com/verdaccio/verdaccio/pull/3370
* [ ] [Harmonize configuration options · Issue #925 · verdaccio/verdaccio · GitHub](https://github.com/verdaccio/verdaccio/issues/925)
* [ ] [Support for multiple access groups in plugins API (access, publish) · Issue #916 · verdaccio/verdaccio · GitHub](https://github.com/verdaccio/verdaccio/issues/916)
* [x] [Allow scopes for loading plugins](https://github.com/verdaccio/verdaccio/issues/1867) #1867 
* [ ] Allow middleware to add layers on different parts of the request (being, end, etc ...) https://github.com/verdaccio/verdaccio/issues/1815#issuecomment-729044946

#### Endpoints

* [x] [Add support for new search endpoint · Issue #310 · verdaccio/verdaccio · GitHub](https://github.com/verdaccio/verdaccio/issues/310) PR https://github.com/verdaccio/verdaccio/pull/1732
* [x] [Feature request: deprecate packages · Issue #1280 · verdaccio/verdaccio · GitHub](https://github.com/verdaccio/verdaccio/issues/1280)
* [x] Enable search v2 https://blog.npmjs.org/post/157615772423/deprecating-the-all-registry-endpoint (there is already an implementation, just make it perfect)

#### Fix Legacy

* [x] [Deprecation Warning: crypto.createDecipher is deprecated. · Issue #1689 · verdaccio/verdaccio · GitHub](https://github.com/verdaccio/verdaccio/issues/1689)
* [ ] [Get rid of async dependency · Issue #1225 · verdaccio/verdaccio · GitHub](https://github.com/verdaccio/verdaccio/issues/1225)
* [x] [Replace bunyan by Pinojs · Issue #1334 · verdaccio/verdaccio · GitHub](https://github.com/verdaccio/verdaccio/issues/1334)
* [ ] Using plugging for custom format logger
* [x] [Allow other password encryptions · Issue #33 · verdaccio/monorepo · GitHub](https://github.com/verdaccio/monorepo/issues/33)
* [ ] [Syntax highlighting in README.md · Issue #245 · verdaccio/monorepo · GitHub](https://github.com/verdaccio/monorepo/issues/245)
* [ ] [refactor: use mkdirSync with the recursive option instead of mkdirp by DanielRuf · Pull Request #1335 · verdaccio/verdaccio · GitHub](https://github.com/verdaccio/verdaccio/pull/1335)
* [x]  Fix ` DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.` message
* [x] feat: remove support for allow_ prefix #1741 
 
#### UI

* [ ] [Link to homepage filtered by keyword · Issue #1226 · verdaccio/verdaccio · GitHub](https://github.com/verdaccio/verdaccio/issues/1226)
* [x] [Configurable UI features · 
  *  Issue #1539 Configurable Login Button
  *  Issue #1720 Hide pnpm instructions on web interface
* [x] ~~[Use custom favicon · Issue #1593 · verdaccio/verdaccio · GitHub](https://github.com/verdaccio/verdaccio/issues/1593) 

#### Project

* [ ] Add Governance file
* [x] Update contributions guidelines
* [ ] Close all `documentation` open at website [Issues · verdaccio/website · GitHub](https://github.com/verdaccio/website/issues?q=is%3Aissue+is%3Aopen+label%3Adocumentation)
* [x] [New Website Design · Issue #75 · verdaccio/website · GitHub](https://github.com/verdaccio/website/issues/75)
  * [ ] Decide whether continous using *Docusaurus* or move to *GatbyJS*
* [x] Build and publish from Docker at GitHub registry 
  * [x] Docker 5.x publish on push (master) #1744 
* [x] Unify repositories under **pnpm** monorepo #1884
  - [x] plugins and core packages from monorepo #1981 #1978 #1977 #1977 #1976 #1975 
- [x] Setup changesets #1913 
- [x] Enable verdaccio standalone app #2046 (run verdaccio with one single JS file, no dependencies) 

#### Migrate from 5.x to 6.x

During the maintenance of verdaccio@5.x some feature were included but not fully ported to the master branch, here the list of PR still pending to be moved which are labeled as `6.x pending`:

- [ ] Pending move 6.x https://github.com/verdaccio/verdaccio/pulls?q=is%3Apr+label%3A%22pending+6.x%22+is%3Aclosed

This is a quite long list of tickets, most of them won't be achieve without your help. Verdaccio maintainers has as firs task guard for stability, help community, bug fixes, releases, promotion and as last in this list improvements. 

> We don't expect any change in the Docker 🐳 image unless any plugin task force to it.

Feel free to comment any of the tickets if you have further questions.

**Comment by crper** [comment]
awesome!

**Comment by dnafication** [comment]
Do we also need to look into fixing the CodeQL errors? Its seems to be complaining about valid issues like hardcoded header values etc.

**Reply by juanpicado** [reply]
CodeQL errors? Could you elaborate?

**Comment by juanpicado** [comment]
No, that's something we are using in a experimental way. Ignore them for now please.

**Comment by ghost** [comment]
Any plans or aspiration to convert the codebase to ES Modules?

**Reply by juanpicado** [reply]
I'd go slow with ES modules
- https://twitter.com/danderson00/status/1365097227916509184
- https://twitter.com/sitnikcode/status/1229806223098306564

~Every time I read a thread about ES modules, people say the ecosystem is not ready. But I'm not 100% into that topic to have a strong opinion yet.~

https://github.com/verdaccio/verdaccio/pull/2985

**Comment by ghost** [comment]
I'm all into ESM on the server side; won't use anything else now. One just needs to know the odd .cjs file extension trick. True that some testing frameworks are still catching up though.

Plus if I can bundle up a Node.JS app into a single .js file, can't see why cannot do same in the browser.

Like all things it took me baby steps to get there but after a few weeks playing around reckon I grok ESM very well now :D

**Reply by juanpicado** [reply]
> Plus if I can bundle up a Node.JS app into a single .js file, can't see why cannot do the same in the browser.

There is an experiment here https://github.com/verdaccio/verdaccio/tree/master/packages/standalone that I was trying to ship all in a single file, but breaks the plugins functionality and does not includes UI.

**Reply by juanpicado** [reply]
> True that some testing frameworks are still catching up though.

Verdaccio is still being used as `require('')` programmatically for so many people that do not have the option to upgrade test runners. But I heard about thr `.cjs` but I am not 100% aware of it, maybe need some reading.
